stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt

```diff
@@ -0,0 +1,797 @@
+import io.getstream.chat.android.offline.plugin.state.global.internal.GlobalMutableState
+import io.getstream.chat.android.offline.repository.builder.internal.RepositoryFacade
+import io.getstream.chat.android.offline.sync.internal.SyncManager
+import io.getstream.logging.StreamLog
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.plus
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import java.util.Date
+import java.util.InputMismatchException
+import java.util.concurrent.atomic.AtomicReference
+
+private const val TAG = "Chat:EventHandlerS"
+private const val TAG_SOCKET = "Chat:SocketEvent"
+private const val EVENTS_BUFFER = 30
+
+/**
+ * Processes events sequentially. That means a new event will not be processed
+ * until the previous event processing is not completed.
+ */
+internal class EventHandlerSequential(
+    scope: CoroutineScope,
+    private val recoveryEnabled: Boolean,
+    private val subscribeForEvents: (ChatEventListener<ChatEvent>) -> Disposable,
+    private val logicRegistry: LogicRegistry,
+    private val stateRegistry: StateRegistry,
+    private val mutableGlobalState: GlobalMutableState,
+    private val repos: RepositoryFacade,
+    private val syncManager: SyncManager,
+) : EventHandler {
+
+    @VisibleForTesting
+    constructor(
+        scope: CoroutineScope,
+        recoveryEnabled: Boolean,
+        subscribeForEvents: (ChatEventListener<ChatEvent>) -> Disposable,
+        logicRegistry: LogicRegistry,
+        stateRegistry: StateRegistry,
+        mutableGlobalState: GlobalMutableState,
+        repos: RepositoryFacade,
+        syncManager: SyncManager,
+        currentUserId: UserId
+    ) : this(
+        scope = scope,
+        recoveryEnabled = recoveryEnabled,
+        subscribeForEvents = subscribeForEvents,
+        logicRegistry = logicRegistry,
+        stateRegistry = stateRegistry,
+        mutableGlobalState = mutableGlobalState,
+        repos = repos,
+        syncManager = syncManager
+    ) {
+        this.currentUserId.set(currentUserId)
+    }
+
+    private val logger = StreamLog.getLogger(TAG)
+    private val scope = scope + SupervisorJob()
+    private val currentUserId = AtomicReference<UserId>()
+    private val socketEvents = MutableSharedFlow<ChatEvent>(extraBufferCapacity = EVENTS_BUFFER)
+    private val socketEventCollector = SocketEventCollector(scope) { pendingBatchEvent ->
+        handleBatchEvent(pendingBatchEvent)
+    }
+
+    private var eventsDisposable: Disposable = EMPTY_DISPOSABLE
+    private var initJob: Job? = null
+
+    override fun initialize(currentUser: User) {
```

Comment:
Should we also introduce `deinitialize` just to make sure `currentUserId ` is cleared and `initJob` joined? 🤔